### DISCUSSION TITLE
Force colour mode as option

### DISCUSSION THREAD

**Discussion by thijsmie** [ewels/rich-click#71] [Q&A] [answered]
I'd like to add an option to my `@click.group` to force rich to use the console colour mode, to enable usecases like `cli --color | less -S -R`. So far I haven't been able to figure out how to manipulate the settings while already in a command (I suspect that is intentionally impossible..). Is this something implementable?

**Comment by ewels** [comment, marked as answer]
Hah, I was thinking / playing with this exact thing last night. You can already sort of force colours by using the `COLOR_SYSTEM` config:

```python
import rich_click as click

click.rich_click.COLOR_SYSTEM = "truecolor"
```

(see the [rich docs](https://rich.readthedocs.io/en/stable/console.html#color-systems) for which systems you can use).

I was also playing with the idea of adding a similar `FORCE_TERMINAL` config option to set [`force_terminal` on the rich console object](https://rich.readthedocs.io/en/stable/console.html#terminal-detection). I'm not entirely sure if that would be all that difference from setting the colour system though 🤔 

Would either of these methods work for you?

**Reply by thijsmie** [reply]
Hi! Thanks for the quick reply! Well, I do know about the `COLOR_SYSTEM` option, however, that does not work when rich has already started its 'Console' object. Consider the following:

```py
import rich_click as click
from rich import print

@click.command()
@click.option('-c', is_flag=True, help="force color mode")
def ls(c):
    if c:
        click.rich_click.COLOR_SYSTEM = "truecolor"
    print("[bold cyan]Hi!")
```

This doesn't work, since the rich `Console` was already created with an automatic color option and it cannot be changed. I cannot detect the value of the flag outside the function without fiddling manually with `sys.argv` again. Maybe then this might need to be solved in Rich itself first, allowing a switch of mode while already running. FYI, I am trying to enable the following usecase:

```shell
# Has colors
ls
# Has no colors, less is not a tty
ls | less
# Has colors, ls is forced into colour mode
ls --color | less -R
```

**Reply by ewels** [reply]
Do you want the output from `--help` coloured or your `print()` statement? The two are separate things - if it's the latter then you can just create your own new `console`:

```python
import rich_click as click
from rich.console import Console

@click.command()
@click.option("-c", is_flag=True, help="force color mode")
def ls(c):
    console = Console(color_system="truecolor") if c else Console()
    console.print("[bold magenta]Hi!")

if __name__ == "__main__":
    ls()

```

If you want to force coloured help text with a flag though (`ls -c --help`) then that's more difficult. Firstly, I'm not sure if it's possible to get flags to alter help text output in click. Secondly, as you correctly pointed out - the help text `console` object is already created before the flags are parsed (I think).

If you want to force coloured help text output I think that we need to use environment variables instead of flags - this is what I was playing with last night. Something like this:

```bash
FORCE_COLOR=1 ls --help | less -R
```

**Reply by thijsmie** [reply]
I can live with that solution, colorized help not working is an edge case I can dismiss. Thanks for the quick help, you'll soon see a PR to [eclipse-cyclonedds/cyclonedds-python](https://github.com/eclipse-cyclonedds/cyclonedds-python) with your useful tool included!

**Reply by ewels** [reply]
Sounds good! I'm currently preparing an update to rich-click with the following addition:

```python
FORCE_TERMINAL = True if getenv("GITHUB_ACTIONS") or getenv("FORCE_COLOR") or getenv("PY_COLORS") else None
```

So the `FORCE_COLOR=1` example above should then work. The behaviour can also be overridden at tool-level by setting `click.rich_click.FORCE_TERMINAL` (just not at cli flag level).

**Reply by ewels** [reply]
ok, added in 6131d55bf7d52f0243064f090e023b9134f170c5 - let me know how you get on!
